File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.js
function solution(board) {
    const dir = [[-1, 0], [0, -1], [1, 0], [0, 1]];
    let robotR, robotC, destR, destC;
    
    board = board.map(row => row.split(""));
    
    for (let r = 0; r < board.length; r++) {
        for (let c = 0; c < board[0].length; c++) {
            if (board[r][c] === "R") {
                robotR = r;
                robotC = c;
            }
            if (board[r][c] === "G") {
                destR = r;
                destC = c;
            }
        }
    }

    const visited = Array.from({length : board.length}, () => Array(board[0].length).fill(false));
    const queue = [[robotR, robotC, 0]];
    visited[robotR][robotC] = true;
    
    while (queue.length > 0) {
        const [curR, curC, count] = queue.shift();
        
        for (let d = 0; d < dir.length; d++) {
            let nr = curR;
            let nc = curC;
            
            while (true) {
                let tr = nr + dir[d][0];
                let tc = nc + dir[d][1];
                
                if (tr < 0 || tc < 0 || tr >= board.length || tc >= board[0].length || board[tr][tc] === 'D') break;
                nr = tr;
                nc = tc;
            }
            
            if (visited[nr][nc]) continue;
            if (nr === destR && nc === destC) return count + 1;
            
            visited[nr][nc] = true;
            queue.push([nr, nc, count + 1]);
        }
    }
    return -1;
}

File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.py
from collections import deque

def solution(board):
    R = len(board)
    C = len(board[0])
    dir = [(-1, 0), (0, -1), (1, 0), (0, 1)]

    visited = [[False] * C for _ in range(R)]
    strR, strC = 0, 0
    endR, endC = 0, 0
    
    for r in range(R):
        for c in range(C):
            if board[r][c] == 'R':
                strR, strC = r, c
            if board[r][c] == 'G':
                endR, endC = r, c
    
    q = deque([(strR, strC, 0)])
    visited[strR][strC] = True
    
    while q:
        r, c, mov = q.popleft()
        
        if r == endR and c == endC:
            return mov
        
        for dr, dc in dir:
            i = 1
            tr, tc = r, c
            while True:
                tr = r + dr * i
                tc = c + dc * i
                if tr < 0 or tr >= R or tc < 0 or tc >= C or board[tr][tc] == 'D':
                    tr, tc = r + dr * (i-1), c + dc * (i-1)
                    break
                i += 1
            nr, nc = tr, tc
            if not visited[nr][nc]:
                visited[nr][nc] = True    
                q.append((nr, nc, mov + 1))
    
    return -1